### PR TITLE
Rewrite outer-joins to inner-join if possible

### PR DESCRIPTION
**Pull request by mfussenegger**

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/symbol/Symbols.java [hunk 1]
```diff
@@ -168,4 +166,38 @@ public Boolean visitReference(Reference symbol, ColumnIdent context) {
...
+        public Symbol visitRelationColumn(RelationColumn relationColumn, RCReplaceCtx ctx) {
+            if (ctx.predicate.apply(relationColumn)) {
+                return ctx.replaceFunction.apply(relationColumn);
+            }
+            return relationColumn;
+        }
+    }
+
+    public static Symbol replaceRC(Symbol symbol,
+                                   Predicate<? super RelationColumn> predicate,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/symbol/Symbols.java]: Might make sense to remove the predicate. The function could do the filtering as well - would remove the need for a separate RCReplaceCtx

**Comment by mfussenegger** [comment]
@dobe please review

File: sql/src/main/java/io/crate/analyze/WhereClause.java [hunk 2]
```diff
@@ -201,4 +202,23 @@ public int hashCode() {
...
+
+
+    /**
+     * Adds another query to this WhereClause.
+     *
+     * The result is either a new WhereClause or the same (but modified) instance.
+     */
+    public WhereClause add(Symbol otherQuery) {
+        assert !docKeys.isPresent() : "Cannot add otherQuery if there are docKeys in the WhereClause";
+        if (this == MATCH_ALL) {
```
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/WhereClause.java]: dunno if checking against the singleton is always enough, probably use the query field?

File: sql/src/main/java/io/crate/analyze/Rewriter.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+                                relationColumn.valueType());
+                        }
+                    }
+                    return input;
+                }
+            });
+        }
+    }
+
+    private static class RelationColumnPredicate implements Predicate<RelationColumn> {
```
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/Rewriter.java]: pls use a more descriptive name

File: sql/src/main/java/io/crate/analyze/symbol/Symbols.java (hunk 1 above)
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/symbol/Symbols.java]: +1

**Comment by dobe** [comment]
if you discover nothing else ... ack

**Pull request merged**